File: backend/app/schemas/pipeline.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class GraphData(BaseModel):
    nodes: list[Any]
    edges: list[Any]


class PipelineCreate(BaseModel):
    name: str
    graph: GraphData


class PipelineUpdate(BaseModel):
    name: str | None = None
    graph: GraphData | None = None


class PipelineResponse(BaseModel):
    pipeline_id: str
    name: str
    graph: Any
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PipelineListItem(BaseModel):
    pipeline_id: str
    name: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/pipeline.ts
export interface GraphNode {
  id: string;
  type: string;
  position: { x: number; y: number };
  data: Record<string, unknown>;
}

export interface GraphEdge {
  id: string;
  source: string;
  sourceHandle: string;
  target: string;
  targetHandle: string;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface Pipeline {
  pipeline_id: string;
  name: string;
  graph: GraphData;
  created_at: string;
  updated_at: string;
}

export interface PipelineListItem {
  pipeline_id: string;
  name: string;
  created_at: string;
  updated_at: string;
}

export interface PipelineCreate {
  name: string;
  graph: GraphData;
}

export interface PipelineUpdate {
  name?: string;
  graph?: GraphData;
}
